File: android/src/com/mku/salmon/vault/dialog/SalmonDialog.java
package com.mku.salmon.vault.dialog;

import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.text.InputType;
import android.view.View;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.mku.fs.drive.utils.FileUtils;
import com.mku.func.Consumer;
import com.mku.salmon.vault.android.R;
import com.mku.salmon.vault.utils.WindowUtils;

import com.mku.func.BiConsumer;

import java.util.TreeMap;

/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
public class SalmonDialog {
    public static void promptEdit(String title, String msg, BiConsumer<String, Boolean> OnEdit,
                                  String value, boolean isFileName, boolean readOnly, boolean isPassword,
                                  String option) {
        Activity activity = WindowUtils.getUiActivity();
        WindowUtils.runOnMainThread(() ->
        {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);

            LinearLayout layout = new LinearLayout(activity);
            layout.setPadding(20, 20, 20, 20);
            layout.setOrientation(LinearLayout.VERTICAL);

            TextInputLayout msgText = new TextInputLayout(activity, null,
                    R.style.Widget_MaterialComponents_TextInputLayout_OutlinedBox);
            msgText.setPadding(20, 20, 20, 20);
            msgText.setBoxBackgroundMode(TextInputLayout.BOX_BACKGROUND_OUTLINE);
            msgText.setBoxCornerRadii(5, 5, 5, 5);
            msgText.setHint(msg);
            View valueText;
            TextInputEditText typePasswd = null;
            if (isPassword) {
                TextInputLayout typePasswdText = new TextInputLayout(activity, null,
                        R.style.Widget_MaterialComponents_TextInputLayout_OutlinedBox);
                typePasswdText.setPasswordVisibilityToggleEnabled(true);
                typePasswdText.setBoxBackgroundMode(TextInputLayout.BOX_BACKGROUND_OUTLINE);
                typePasswdText.setBoxCornerRadii(5, 5, 5, 5);
                typePasswdText.setHint(msg);

                typePasswd = new TextInputEditText(typePasswdText.getContext());
                typePasswd.setInputType(InputType.TYPE_TEXT_VARIATION_PASSWORD |
                        InputType.TYPE_CLASS_TEXT);
                typePasswd.setText(value);
                typePasswdText.addView(typePasswd);
                valueText = typePasswdText;
            } else {
                TextInputEditText text = new TextInputEditText(msgText.getContext());
                text.setText(value);
                if (readOnly) {
                    text.setInputType(InputType.TYPE_NULL);
                    text.setTextIsSelectable(true);
                } else {
                    text.setInputType(InputType.TYPE_CLASS_TEXT);
                }
                text.setOnFocusChangeListener(new View.OnFocusChangeListener() {
                    boolean once = false;

                    @Override
                    public void onFocusChange(View view, boolean b) {
                        if (!once) {
                            if (isFileName) {
                                String ext = FileUtils.getExtensionFromFileName(value);
                                if (ext != null && ext.length() > 0)
                                    text.setSelection(0, value.length() - ext.length() - 1);
                                else
                                    text.setSelection(0, value.length());
                            } else {
                                text.selectAll();
                            }
                            once = true;
                        }
                    }
                });
                valueText = text;
            }
            LinearLayout.LayoutParams parameters = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT);
            msgText.addView(valueText);
            layout.addView(msgText, parameters);
            CheckBox optionCheckBox = new CheckBox(activity);
            if (option != null) {
                optionCheckBox.setText(option);
                layout.addView(optionCheckBox, parameters);
            }
            TextInputEditText finalTypePasswd = typePasswd;
            builder.setPositiveButton(activity.getString(android.R.string.ok), (sender, e) ->
            {
                if (OnEdit != null) {
                    if (isPassword)
                        OnEdit.accept(finalTypePasswd.getText().toString(), optionCheckBox.isChecked());
                    else
                        OnEdit.accept(((EditText) valueText).getText().toString(), optionCheckBox.isChecked());
                }
            });
            builder.setNegativeButton(activity.getString(android.R.string.cancel), (sender, e) ->
            {
                ((AlertDialog) sender).dismiss();
            });

            AlertDialog alertDialog = builder.create();
            alertDialog.setTitle(title);
            alertDialog.setCancelable(true);
            alertDialog.setCanceledOnTouchOutside(false);
            alertDialog.setView(layout);

            valueText.requestFocus();
            alertDialog.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
            if (!activity.isFinishing())
                alertDialog.show();

        });
    }


    public static void promptCredentialsEdit(String title, String msg,
                                             String[] hints, String[] values, boolean[] isPasswords,
                                             Consumer<String[]> OnEdit) {
        Activity activity = WindowUtils.getUiActivity();
        WindowUtils.runOnMainThread(() -> {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);

            LinearLayout layout = new LinearLayout(activity);
            layout.setPadding(20, 20, 20, 20);
            layout.setOrientation(LinearLayout.VERTICAL);

            LinearLayout.LayoutParams parameters = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT);

            TextView msgText = new TextView(activity);
            msgText.setPadding(20, 20, 20, 20);
            msgText.setText(msg);
            layout.addView(msgText, parameters);

            View[] textFields = new View[hints.length];
            for (int i = 0; i < hints.length; i++) {
                if (hints[i] != null) {
                    textFields[i] = createTextField(activity, hints[i], values[i], isPasswords[i]);
                    layout.addView(textFields[i], parameters);
                }
            }

            builder.setPositiveButton(activity.getString(android.R.string.ok), (sender, e) ->
            {
                if (OnEdit != null) {
                    String[] texts = new String[hints.length];
                    for (int i = 0; i < textFields.length; i++) {
                        if (textFields[i] != null) {
                            EditText editText;
                            if(textFields[i] instanceof TextInputLayout)
                                editText = ((TextInputLayout) textFields[i]).getEditText();
                            else
                                editText = (EditText) textFields[i];
                            texts[i] = editText.getText().toString();
                        }
                    }
                    OnEdit.accept(texts);
                }
            });
            builder.setNegativeButton(activity.getString(android.R.string.cancel), (sender, e) ->
            {
                sender.dismiss();
            });

            AlertDialog alertDialog = builder.create();
            alertDialog.setTitle(title);
            alertDialog.setCancelable(true);
            alertDialog.setCanceledOnTouchOutside(false);
            alertDialog.setView(layout);

            alertDialog.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
            if (!activity.isFinishing())
                alertDialog.show();
        });
    }

    private static View createTextField(Activity activity, String hint, String value, boolean isPassword) {

        View valueText;
        if(!isPassword) {
            TextInputEditText text = new TextInputEditText(activity);
            text.setInputType(InputType.TYPE_CLASS_TEXT);
            text.setHint(hint);
            text.setText(value);
            valueText = text;
        } else {
            TextInputLayout typePasswdText = new TextInputLayout(activity, null,
                    R.style.Widget_MaterialComponents_TextInputLayout_OutlinedBox);
            typePasswdText.setPasswordVisibilityToggleEnabled(true);
            typePasswdText.setBoxBackgroundMode(TextInputLayout.BOX_BACKGROUND_OUTLINE);
            typePasswdText.setBoxCornerRadii(5, 5, 5, 5);
            typePasswdText.setHint(hint);

            TextInputEditText typePasswd = new TextInputEditText(typePasswdText.getContext());
            typePasswd.setInputType(InputType.TYPE_TEXT_VARIATION_PASSWORD |
                    InputType.TYPE_CLASS_TEXT);
            typePasswd.setText(value);
            typePasswdText.addView(typePasswd);
            valueText = typePasswdText;
        }
        return valueText;
    }

    public static void promptDialog(String title, String body,
                                    String buttonLabel1, Runnable buttonListener1,
                                    String buttonLabel2, Runnable buttonListener2) {
        promptDialog(title, body, buttonLabel1, buttonListener1, buttonLabel2, buttonListener2,
        null, null);
    }

    public static void promptDialog(String title, String body,
                                    String buttonLabel1, Runnable buttonListener1,
                                    String buttonLabel2, Runnable buttonListener2,
                                    String buttonLabel3, Runnable buttonListener3) {
        if (buttonLabel1 == null)
            buttonLabel1 = "Ok";
        Activity activity = WindowUtils.getUiActivity();
        String finalButtonLabel = buttonLabel1;
        WindowUtils.runOnMainThread(() ->
        {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
            if (title != null)
                builder.setTitle(title);
            builder.setMessage(body);
            if (finalButtonLabel != null)
                builder.setPositiveButton(finalButtonLabel, (s, e) ->
                {
                    if (buttonListener1 != null)
                        buttonListener1.run();
                });
            if (buttonLabel2 != null)
                builder.setNegativeButton(buttonLabel2, (s, e) ->
                {
                    if (buttonListener2 != null)
                        buttonListener2.run();
                });
            if (buttonLabel3 != null)
                builder.setNeutralButton(buttonLabel3, (s, e) ->
                {
                    if (buttonListener3 != null)
                        buttonListener3.run();
                });

            AlertDialog alertDialog = builder.create();
            alertDialog.setTitle(title);
            alertDialog.setCancelable(true);
            alertDialog.setCanceledOnTouchOutside(false);

            if (!activity.isFinishing())
                alertDialog.show();
        });
    }

    public static Consumer<String> promptUpdatableDialog(String title, String msg) {
        Activity activity = WindowUtils.getUiActivity();
        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);

        LinearLayout layout = new LinearLayout(activity);
        layout.setPadding(20, 20, 20, 20);
        TextView textView = new TextView(activity);
        textView.setPadding(20, 20, 20, 20);
		textView.setText(msg);
        layout.addView(textView);
        builder.setPositiveButton(android.R.string.ok, null);

        AlertDialog alertDialog = builder.create();
        alertDialog.setTitle(title);
        alertDialog.setCancelable(true);
        alertDialog.setCanceledOnTouchOutside(false);
        alertDialog.setView(layout);
        if (!activity.isFinishing())
            alertDialog.show();
        return (body) -> WindowUtils.runOnMainThread(() -> textView.setText(body));
    }

    public static void promptSingleValue(ArrayAdapter<String> adapter, String title,
                                         int currSelection, BiConsumer<AlertDialog, Integer> onClickListener) {
        Activity activity = WindowUtils.getUiActivity();
        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
        if (title != null)
            builder.setTitle(title);
        builder.setSingleChoiceItems(adapter, currSelection, (sender, which) ->
        {
            onClickListener.accept((AlertDialog) sender, which);
        });
        AlertDialog alertDialog = builder.create();
        alertDialog.setTitle(title);
        alertDialog.setCancelable(true);
        alertDialog.setCanceledOnTouchOutside(false);

        if (!activity.isFinishing())
            alertDialog.show();
    }

    public static void promptDialog(String body) {
        promptDialog("", body, null, null, null, null);
    }

    public static void promptDialog(String title, String body) {
        promptDialog(title, body, null, null, null, null);
    }

    public static void promptDialog(String title, String body, String buttonLabel) {
        promptDialog(title, body, buttonLabel, null, null, null);
    }
}

File: dotnet-android/src/Dialog/SalmonDialog.cs
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
using Android.Content;
using Android.Widget;
using Google.Android.Material.Dialog;
using Google.Android.Material.TextField;
using Salmon.Vault.Utils;
using System;
using AndroidX.AppCompat.App;
using Android.Views;
using Activity = Android.App.Activity;
using Salmon.Vault.DotNetAndroid;
using Mku.FS.Drive.Utils;
using System.Collections.Generic;
using static Salmon.Vault.Main.SalmonActivity;
using System.Linq;
using static Android.Provider.CalendarContract;

namespace Salmon.Vault.Dialog;

public class SalmonDialog
{
    public static void PromptEdit(string title, string msg, Action<string, bool> OnEdit,
        string value = "", bool isFileName = false, bool readOnly = false, bool isPassword = false,
        string option = null)
    {
        Activity activity = WindowUtils.UiActivity;
        WindowUtils.RunOnMainThread(() =>
        {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);

            LinearLayout layout = new LinearLayout(activity);
            layout.SetPadding(20, 20, 20, 20);
            layout.Orientation = Orientation.Vertical;

            TextInputLayout msgText = new TextInputLayout(activity, null,
                Resource.Style.Widget_MaterialComponents_TextInputLayout_OutlinedBox);
            msgText.SetPadding(20, 20, 20, 20);
            msgText.BoxBackgroundMode = TextInputLayout.BoxBackgroundOutline;
            msgText.SetBoxCornerRadii(5, 5, 5, 5);
            msgText.Hint = msg;
            View valueText;
            TextInputEditText typePasswd = null;
            if (isPassword)
            {
                TextInputLayout typePasswdText = new TextInputLayout(activity, null,
            Resource.Style.Widget_MaterialComponents_TextInputLayout_OutlinedBox);
                typePasswdText.PasswordVisibilityToggleEnabled = true;
                typePasswdText.BoxBackgroundMode = TextInputLayout.BoxBackgroundOutline;
                typePasswdText.SetBoxCornerRadii(5, 5, 5, 5);
                typePasswdText.Hint = msg;

                typePasswd = new TextInputEditText(typePasswdText.Context);
                typePasswd.InputType = Android.Text.InputTypes.TextVariationPassword |
                              Android.Text.InputTypes.ClassText;
                typePasswd.Text = value;
                typePasswdText.AddView(typePasswd);
                valueText = typePasswdText;
            }
            else
            {
                TextInputEditText text = new TextInputEditText(msgText.Context);
                text.Text = value;
                if (readOnly)
                {
                    text.InputType = Android.Text.InputTypes.Null;
                    text.SetTextIsSelectable(true);
                }
                else
                {
                    text.InputType = Android.Text.InputTypes.ClassText;
                }
                bool once = false;
                text.FocusChange += (sender, args) =>
                {
                    if (!once)
                    {
                        if (isFileName)
                        {
                            string ext = FileUtils.GetExtensionFromFileName(value);
                            if (ext != null && ext.Length > 0)
                                text.SetSelection(0, value.Length - ext.Length - 1);
                            else
                                text.SetSelection(0, value.Length);
                        }
                        else
                        {
                            text.SelectAll();
                        }
                        once = true;
                    }
                };
                valueText = text;
            }
            LinearLayout.LayoutParams parameters = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MatchParent, LinearLayout.LayoutParams.MatchParent);
            msgText.AddView(valueText);
            layout.AddView(msgText, parameters);
            CheckBox optionCheckBox = new CheckBox(activity);
            if (option != null)
            {
                optionCheckBox.Text = option;
                layout.AddView(optionCheckBox, parameters);
            }
            builder.SetPositiveButton(activity.GetString(Android.Resource.String.Ok), (object sender, DialogClickEventArgs e) =>
            {
                if (OnEdit != null)
                {
                    if (isPassword)
                        OnEdit(typePasswd.Text.ToString(), optionCheckBox.Checked);
                    else
                        OnEdit((valueText as EditText).Text.ToString(), optionCheckBox.Checked);
                }
            });
            builder.SetNegativeButton(activity.GetString(Android.Resource.String.Cancel), (object sender, DialogClickEventArgs e) =>
            {
                ((AlertDialog)sender).Dismiss();
            });

            AlertDialog alertDialog = builder.Create();
            alertDialog.SetTitle(title);
            alertDialog.SetCancelable(true);
            alertDialog.SetCanceledOnTouchOutside(false);
            alertDialog.SetView(layout);

            valueText.RequestFocus();
            alertDialog.Window.SetSoftInputMode(SoftInput.AdjustResize);
            if (!activity.IsFinishing)
                alertDialog.Show();
        });
    }


    public static void PromptCredentialsEdit(string title, String msg,
                                             string[] hints, String[] values, bool[] isPasswords,
                                             Action<string[]> OnEdit)
    {
        Activity activity = WindowUtils.UiActivity;
        WindowUtils.RunOnMainThread(() =>
        {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);

            LinearLayout layout = new LinearLayout(activity);
            layout.SetPadding(20, 20, 20, 20);
            layout.Orientation = Orientation.Vertical;

            LinearLayout.LayoutParams parameters = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MatchParent, LinearLayout.LayoutParams.MatchParent);

            TextView msgText = new TextView(activity);
            msgText.SetPadding(20, 20, 20, 20);
            msgText.Text = msg;
            layout.AddView(msgText, parameters);

            View[] textFields = new View[hints.Length];
            for (int i = 0; i < hints.Length; i++)
            {
                if (hints[i] != null)
                {
                    textFields[i] = CreateTextField(activity, hints[i], values[i], isPasswords[i]);
                    layout.AddView(textFields[i], parameters);
                }
            }

            builder.SetPositiveButton(activity.GetString(Android.Resource.String.Ok), (sender, e) =>
            {
                if (OnEdit != null)
                {
                    String[] texts = new String[hints.Length];
                    for (int i = 0; i < textFields.Length; i++)
                    {
                        if (textFields[i] != null)
                        {
                            EditText editText;
                            if (textFields[i] is TextInputLayout)
                                editText = ((TextInputLayout)textFields[i]).EditText;
                            else
                                editText = (EditText)textFields[i];
                            texts[i] = editText.Text.ToString();
                        }
                    }
                    OnEdit(texts);
                }
            });
            builder.SetNegativeButton(activity.GetString(Android.Resource.String.Cancel), (sender, e) =>
            {
                ((AlertDialog)sender).Dismiss();
            });

            AlertDialog alertDialog = builder.Create();
            alertDialog.SetTitle(title);
            alertDialog.SetCancelable(true);
            alertDialog.SetCanceledOnTouchOutside(false);
            alertDialog.SetView(layout);

            alertDialog.Window.SetSoftInputMode(SoftInput.AdjustResize);
            if (!activity.IsFinishing)
                alertDialog.Show();
        });
    }

    private static View CreateTextField(Activity activity, String hint, String value, bool isPassword)
    {

        View valueText;
        if (!isPassword)
        {
            TextInputEditText text = new TextInputEditText(activity);
            text.InputType = Android.Text.InputTypes.ClassText;
            text.Hint = hint;
            text.Text = value;
            valueText = text;
        }
        else
        {
            TextInputLayout typePasswdText = new TextInputLayout(activity, null,
                    Resource.Style.Widget_MaterialComponents_TextInputLayout_OutlinedBox);
            typePasswdText.PasswordVisibilityToggleEnabled = true;
            typePasswdText.BoxBackgroundMode = TextInputLayout.BoxBackgroundOutline;
            typePasswdText.SetBoxCornerRadii(5, 5, 5, 5);
            typePasswdText.Hint = hint;

            TextInputEditText typePasswd = new TextInputEditText(typePasswdText.Context);
            typePasswd.InputType = Android.Text.InputTypes.TextVariationPassword |
                    Android.Text.InputTypes.ClassText;
            typePasswd.Text = value;
            typePasswdText.AddView(typePasswd);
            valueText = typePasswdText;
        }
        return valueText;
    }

    public static void PromptDialog(string title, string body,
                                        string buttonLabel1 = "Ok", Action buttonListener1 = null,
                                        string buttonLabel2 = null, Action buttonListener2 = null)
    {
        Activity activity = WindowUtils.UiActivity;
        WindowUtils.RunOnMainThread(() =>
        {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
            if (title != null)
                builder.SetTitle(title);
            builder.SetMessage(body);
            if (buttonLabel1 != null)
                builder.SetPositiveButton(buttonLabel1, (s, e) =>
                {
                    if (buttonListener1 != null)
                        buttonListener1();
                });
            if (buttonLabel2 != null)
                builder.SetNegativeButton(buttonLabel2, (s, e) =>
                {
                    if (buttonListener2 != null)
                        buttonListener2();
                });

            AlertDialog alertDialog = builder.Create();
            alertDialog.SetTitle(title);
            alertDialog.SetCancelable(true);
            alertDialog.SetCanceledOnTouchOutside(false);

            if (!activity.IsFinishing)
                alertDialog.Show();
        });
    }

    public static Action<string> promptUpdatableDialog(string title, string msg)
    {
        Activity activity = WindowUtils.UiActivity;
        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);

        LinearLayout layout = new LinearLayout(activity);
        layout.SetPadding(20, 20, 20, 20);
        TextView textView = new TextView(activity);
        textView.SetPadding(20, 20, 20, 20);
        textView.Text = msg;
        layout.AddView(textView);
        builder.SetPositiveButton(Android.Resource.String.Ok, (sender, e) => { });

        AlertDialog alertDialog = builder.Create();
        alertDialog.SetTitle(title);
        alertDialog.SetCancelable(true);
        alertDialog.SetCanceledOnTouchOutside(false);
        alertDialog.SetView(layout);
        if (!activity.IsFinishing)
            alertDialog.Show();
        return (body) => WindowUtils.RunOnMainThread(() => textView.Text = body);
    }

    public static void PromptSingleValue(string title, IList<string> items,
                                         int currSelection, Action<int> onClickListener)
    {
        Activity activity = WindowUtils.UiActivity;
        ArrayAdapter<string> adapter = new ArrayAdapter<string>(
                activity, Android.Resource.Layout.SimpleListItemActivated1, items.ToArray());
        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
        if (title != null)
            builder.SetTitle(title);
        builder.SetSingleChoiceItems(adapter, currSelection, (object sender, DialogClickEventArgs e) =>
        {
            ((AlertDialog)sender).Dismiss();
            onClickListener(e.Which);
        });
        AlertDialog alertDialog = builder.Create();
        alertDialog.SetTitle(title);
        alertDialog.SetCancelable(true);
        alertDialog.SetCanceledOnTouchOutside(false);
        if (!activity.IsFinishing)
            alertDialog.Show();
    }

    public static void PromptDialog(string body)
    {
        WindowUtils.RunOnMainThread(() =>
        {
            PromptDialog("", body, "Ok");
        });
    }
}